### ISSUE TITLE
Incapsula analysis progress

### ISSUE THREAD

**Issue by rbri** [HtmlUnit/htmlunit#552]
HtmlUnit fails if you try to get web pages hosted by / protected by the incapsula script.
Because we have some reports about this i try to make another round of figuring out what goes wrong here.

This is about the different findings....

**Comment by rbri** [comment]
Problem 1 - the outer obfuscating

The script itself looks like this

```javascript
(function() {
    var z="";
    var b="76617220........";
    eval((function() {
        for (var i=0;i<b.length;i+=2) {
             z+=String.fromCharCode(parseInt(b.substring(i,i+2),16));
        };
        return z;
    })());
})();
```

This simple encoding invokes really many string concatenations - and in Rhino there is a special optimization (see ConsString) that results in really the opposite of optimization for this case.

To overcome this problem i wrote a small connection wrapper that does this in java instead in javascript:

```java
    new WebConnectionWrapper(webClient) {
        public WebResponse getResponse(WebRequest request) throws IOException {
            webResponse response = super.getResponse(request);
            if (request.getUrl().toExternalForm().contains("Incapsula")) {
                String js = response.getContentAsString();
                js = js.substring(js.indexOf("b=\"") + 3, js.indexOf(";eval"));

                StringBuilder res = new StringBuilder();
                for (int i = 0; i < js.length()-1; i += 2) {
                    String str = js.substring(i, i + 2);
                    res.append((char) Integer.parseInt(str, 16));
                }

                // content is no longer gzip encoded
                List<NameValuePair> header = response.getResponseHeaders();
                List<NameValuePair> cleanedHeader = new ArrayList<>();
                for (NameValuePair pair : header) {
                    if (pair.getName().equalsIgnoreCase("Content-Encoding")) {
                        continue;
                    }
                    cleanedHeader.add(pair);
                }
                WebResponseData data = new WebResponseData(res.toString().getBytes(response.getContentCharset()),
                                                         response.getStatusCode(), response.getStatusMessage(), cleanedHeader);
                response = new WebResponse(data, request, response.getLoadTime());
            }
            return response;
        }
    };
```

**Comment by rbri** [comment]
The result looks like this

    var _0xe285=['\x55\x6d\x2f\x44\x69\x51\x3d\x3d','\x77.....

means it is still obfuscated.

**Comment by rbri** [comment]
Problem 2
The code has some protection against de obfuscating - this  works based on
* using ToString() on functions to check if the function source is still obfuscated (and still in one line)
* if the function code looks changed the code runs into an endless loop (at least into a code part that fails after some time in real browsers by increasing a string until the browser reaches the string limit - maybe to avoid DoS attacks.

Sample

```javascript
// fake function only for checking later
'removeCookie': function(){return'\x64\x65\x76';}

// this function checks if the removeCookie function source is unchanged
var _0x447b20 = function() {
    var _0x1e5fc4 = new RegExp('\\w+ *\\(\\) *{\\w+ *[\'|\"].+[\'|\"];? *}');
    return _0x1e5fc4['test'](_0x1e0aa6['removeCookie']['toString']());
};

// the endless loop
// the upper border of the loop condition _0x2c98b5 always updated inside the loop
'setCookie': function(_0x47572c, _0x2e4c5c, _0x3f19cc, _0x26974f) {
    _0x26974f = _0x26974f || {};
    var _0x210528 = _0x2e4c5c + '=' + _0x3f19cc;
    var _0x31ef31 = 0x0;
    for (var _0x31ef31 = 0x0, _0x2c98b5 = _0x47572c['length']; _0x31ef31 < _0x2c98b5; _0x31ef31++) {
        var _0x4356d9 = _0x47572c[_0x31ef31];
        _0x210528 += '; ' + _0x4356d9;
        var _0x203e25 = _0x47572c[_0x4356d9];
        _0x47572c['push'](_0x203e25);
        _0x2c98b5 = _0x47572c['length'];
        if (_0x203e25 !== !![]) {
            _0x210528 += '=' + _0x203e25;
        }
    }
    _0x26974f['cookie'] = _0x210528;
},

// finally the check this calls the check function _0x447b20 and if the check fails
// the setCookie-endless loop is called 
_0x1e0aa6['updateCookie'] = _0x447b20;
var _0x551e6f = '';
var _0x151302 = _0x1e0aa6['updateCookie']();
if (!_0x151302) {
    _0x1e0aa6['setCookie'](['*'], 'counter', 0x1);
} else if (_0x151302) {
    _0x551e6f = _0x1e0aa6['getCookie'](null, 'counter');
} else {
    _0x1e0aa6['removeCookie']();
}
```

There are several checks like this inside the code

**Comment by rbri** [comment]
As there are different checks working based on toString() of functions we need a real fix for the Function.toString() implementation. Rhino uses a serialization of the AST and decompilation (in fact deserialization) on request. This has to be replaces by storing the source code instead.

**Comment by rbri** [comment]
After a bit more analysis - there is no real problem with the string performance; the WebConnectionWrapper is not required. Doing the 'decoding' of the script with core-js is nearly as fast as the StringBuilder based WebConnectionWrapper.

**Comment by rbri** [comment]
released with version 3.0.0
